src/Processors/QueryPlan/Optimizations/optimizeJoin.cpp

```diff
@@ -159,247 +322,731 @@ bool optimizeJoinLegacy(QueryPlan::Node & node, QueryPlan::Nodes &, const QueryP
+    std::unordered_map<BitSet, String> relation_names;
+    for (size_t i = 0; i < query_graph.relation_stats.size(); ++i)
     {
-        StatsCollectingParams params{
-            /*key_=*/0,
-            /*enable=*/true,
-            optimization_settings.max_entries_for_hash_table_stats,
-            optimization_settings.max_size_to_preallocate_for_joins};
-        if (auto hint = getHashTablesStatistics<HashJoinEntry>().getSizeHint(params.setKey(hash_table_key_hashes->key_hash_left)))
-            lhs_estimation = std::min<size_t>(lhs_estimation.value_or(std::numeric_limits<size_t>::max()), hint->source_rows);
-        if (auto hint = getHashTablesStatistics<HashJoinEntry>().getSizeHint(params.setKey(hash_table_key_hashes->key_hash_right)))
-            rhs_estimation = std::min<size_t>(rhs_estimation.value_or(std::numeric_limits<size_t>::max()), hint->source_rows);
+        const auto & table_name = query_graph.relation_stats[i].table_name;
+        auto estimated_count = query_graph.relation_stats[i].estimated_rows;
+        String estimation = estimated_count ? fmt::format("[{}]", estimated_count.value()) : "";
+        if (!table_name.empty())
+            relation_names[BitSet().set(i)] = fmt::format("{}{}", table_name, estimation);
+        else
+            relation_names[BitSet().set(i)] = fmt::format("R{}{}", i, estimation);
     }
 
-    LOG_TRACE(
-        getLogger("optimizeJoin"),
-        "Left table estimation: {}, right table estimation: {}",
-        lhs_estimation.transform(toString<UInt64>).value_or("unknown"),
-        rhs_estimation.transform(toString<UInt64>).value_or("unknown"));
+    auto global_expression_actions = std::move(query_graph_builder.expression_actions);
+    auto global_actions_dag = global_expression_actions.getActionsDAG();
+    if (!global_actions_dag)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Global expression actions DAG is not set");
+
+    auto optimized = optimizeJoinOrder(std::move(query_graph));
+    auto sequence = getJoinTreePostOrderSequence(optimized);
+
+    if (sequence.empty())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Join tree is empty");
+
+    std::unordered_map<const ActionsDAG::Node *, size_t> usage_level_map;
 
-    if (!optimization_settings.join_swap_table.has_value())
+    for (size_t i = 0; i < sequence.size(); ++i)
     {
-        if (lhs_estimation && rhs_estimation && *lhs_estimation < *rhs_estimation)
-            need_swap = true;
+        const auto & entry = sequence[i];
+        for (const auto & action : entry->join_operator.expression)
+            usage_level_map[action.getNode()] = i;
+        for (const auto & action : entry->join_operator.residual_filter)
+            usage_level_map[action.getNode()] = i;
     }
-    else if (optimization_settings.join_swap_table.value())
+
+    for (const auto & out : global_actions_dag->getOutputs())
+        usage_level_map[out] = sequence.size();
```

Comment:
Can we override the previous value from expression/residual filter? E.g. in `select l.x from l join r on l.x = r.x and l.x`?

Reply:
Yes, I think so, it's intended. We need to take latest stage where node is used.